In [46]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from nltk import word_tokenize
from nltk.tokenize import wordpunct_tokenize
from nltk.stem.snowball import EnglishStemmer
from nltk import RegexpTokenizer

from nltk.corpus import stopwords

from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer, TfidfVectorizer
from sklearn.dummy import DummyClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report, f1_score, accuracy_score, confusion_matrix
from sklearn.pipeline import Pipeline
from sklearn.model_selection import StratifiedKFold, cross_val_score, train_test_split

In [47]:
# Из наброска кода
messages = pd.read_csv('SMSSpamCollection', sep='\t', names=["label", "message"])
messages

label                                            message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
5     spam  FreeMsg Hey there darling it's been 3 week's n...
6      ham  Even my brother is not like to speak with me. ...
7      ham  As per your request 'Melle Melle (Oru Minnamin...
8     spam  WINNER!! As a valued network customer you have...
9     spam  Had your mobile 11 months or more? U R entitle...
10     ham  I'm gonna be home soon and i don't want to tal...
11    spam  SIX chances to win CASH! From 100 to 20,000 po...
12    spam  URGENT! You have won a 1 week FREE membership ...
13     ham  I've been searching for the right words to tha...
14     ham                I HAVE A DATE ON SUNDAY WITH WILL!!
15    spam  XXXMobileMovieClub: To use your credit, click ...
16     ham                         Oh k...i'm watching here:)
17     ham  Eh u remember how 2 spell his name... Yes i di...
18     ham  Fine if that’s the way u feel. That’s the way ...
19    spam  England v Macedonia - dont miss the goals/team...
20     ham          Is that seriously how you spell his name?
21     ham    I‘m going to try for 2 months ha ha only joking
22     ham  So ü pay first lar... Then when is da stock co...
23     ham  Aft i finish my lunch then i go str down lor. ...
24     ham  Ffffffffff. Alright no way I can meet up with ...
25     ham  Just forced myself to eat a slice. I'm really ...
26     ham                     Lol your always so convincing.
27     ham  Did you catch the bus ? Are you frying an egg ...
28     ham  I'm back &amp; we're packing the car now, I'll...
29     ham  Ahhh. Work. I vaguely remember that! What does...
...    ...                                                ...
5542   ham           Armand says get your ass over to epsilon
5543   ham             U still havent got urself a jacket ah?
5544   ham  I'm taking derek &amp; taylor to walmart, if I...
5545   ham      Hi its in durban are you still on this number
5546   ham         Ic. There are a lotta childporn cars then.
5547  spam  Had your contract mobile 11 Mnths? Latest Moto...
5548   ham                 No, I was trying it all weekend ;V
5549   ham  You know, wot people wear. T shirts, jumpers, ...
5550   ham        Cool, what time you think you can get here?
5551   ham  Wen did you get so spiritual and deep. That's ...
5552   ham  Have a safe trip to Nigeria. Wish you happines...
5553   ham                        Hahaha..use your brain dear
5554   ham  Well keep in mind I've only got enough gas for...
5555   ham  Yeh. Indians was nice. Tho it did kane me off ...
5556   ham  Yes i have. So that's why u texted. Pshew...mi...
5557   ham  No. I meant the calculation is the same. That ...
5558   ham                             Sorry, I'll call later
5559   ham  if you aren't here in the next  &lt;#&gt;  hou...
5560   ham                  Anything lor. Juz both of us lor.
5561   ham  Get me out of this dump heap. My mom decided t...
5562   ham  Ok lor... Sony ericsson salesman... I ask shuh...
5563   ham                                Ard 6 like dat lor.
5564   ham  Why don't you wait 'til at least wednesday to ...
5565   ham                                       Huh y lei...
5566  spam  REMINDER FROM O2: To get 2.50 pounds free call...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

Задание 1(+6). Проверить, сбалансирован ли датасет (может быть, наблюдений одного класса слишком много?). Какие результаты покажет dummy classifier, который будет всем новым наблюдениям присваивать класс ham? Насколько плохо такое решение для задачи определения спама?
Грубое решение - включить в training set только необходимое число наблюдений (примерно поровну spam и ham).

Нормализовать тексты и обучить байесовскую модель (bag of words). Проверить, как влияют на результат:

1) разная токенизация: в одном случае знаки препинания удалять, в другом — считать их токенами; 

2) лемматизация (отсутствие лемматизации, стемминг, лемматизация; инструменты можно использовать любые, например, nltk.stem); 

3) удаление стоп-слов, а также пороги минимальной и максимальной document frequency;

4) векторизация документов (CountVectorizer vs. TfIdfVectorizer);

5) что-нибудь ещё?

При оценке классификатора обратите внимание на TP и FP.

In [48]:
# Из наброска кода, тааакс посмотрим на статистичку
messages.groupby('label').describe()

message
label                                                          
ham   count                                                4825
      unique                                               4516
      top                                Sorry, I'll call later
      freq                                                   30
spam  count                                                 747
      unique                                                653
      top     Please call our customer service representativ...
      freq                                                    4

Отсюда мы видим, что выборка несбалансирована - спама гораздо меньше (747), чем хама (4825). В такои случае мы должны уравнять спам и хам по количеству (хоть это и грубое решение, я что-то не могу придумать, как это можно было бы сделать по-другому, иначе результат будет нерелевантен).

In [49]:
spam = messages[messages['label'] == 'spam']
ham = messages[messages['label'] == 'ham'][:len(spam)]
messages = pd.concat([spam, ham])
messages['length'] = messages['message'].map(lambda text: len(text))

In [50]:
# Функция для токенизации, удаляющей знаки препинания. Тоже из наброска код
def tokenize(text):
    text = text.lower()
    return word_tokenize(text)
#messages.message.head().apply(tokenize_punc)
#messages.message = messages.message.apply(tokenize_punc)
#print(messages.head())

Сразу заметно, что в спаме используется очень много !!

In [51]:
# Функция для токенизации, сохраняющей знаки препинания.
def tokenize_punc(text):
    text = text.lower()
    return wordpunct_tokenize(text)
#messages.message.head().apply(tokenize_punc)
#messages.message = messages.message.apply(tokenize_punc)
#print(messages.head())

In [52]:
# Стемматизация
def tokenize_stem(text):
    stems = []
    for i in RegexpTokenizer(r'\w+').tokenize(text):
        stems.append(EnglishStemmer(ignore_stopwords=True).stem(i))            
    return stems

In [53]:
# Стоп-слова подкатили
stopwords = stopwords.words('english')

In [54]:
# Векторизуем по CountVectorizer (часть кода из наброска) и смотрим на DummyClassifier
bow = CountVectorizer()
bow.fit_transform(messages['message'])

bowed_messages = bow.transform(messages['message'])
dummy_model = DummyClassifier(random_state=0, strategy='most_frequent')
dummy_model.fit(bowed_messages, messages['label'])
print(classification_report(messages['label'], dummy_model.predict(bowed_messages)))

             precision    recall  f1-score   support

        ham       0.50      1.00      0.67       747
       spam       0.00      0.00      0.00       747

avg / total       0.25      0.50      0.33      1494



/Users/NP/anaconda/lib/python3.6/site-packages/sklearn/metrics/classification.py:1113: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [55]:
# Далаем тоже самое только с TfidfVectorizer, смотрим на DummyClassifier в этом случае. Может быть, это влияет на результат
bow = TfidfVectorizer()
bow.fit_transform(messages['message'])

b_messages = bow.transform(messages['message'])
dummy_model = DummyClassifier(random_state=0, strategy='most_frequent')
dummy_model.fit(b_messages, messages['label'])
print(classification_report(messages['label'], dummy_model.predict(bowed_messages)))

             precision    recall  f1-score   support

        ham       0.50      1.00      0.67       747
       spam       0.00      0.00      0.00       747

avg / total       0.25      0.50      0.33      1494



/Users/NP/anaconda/lib/python3.6/site-packages/sklearn/metrics/classification.py:1113: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


Нет, на результат это не повлияло. Но выводы по DummyClassifier сделать можно. Его точно нельзя применять в этом случае. Нам же нужно иденфицировать спам сообщения, он их не определяет вообще (f1 score, precision, recall - всё по нулям).

Теперь посмотрим на Байескую модель и как на неё влияют разные условия

In [34]:
# CountVectorizer и токенизация с удалением знаков
bow = CountVectorizer(analyzer=tokenize)
bow.fit_transform(messages['message'])
bowed_messages = bow.transform(messages['message'])
naive_model = MultinomialNB()
naive_model.fit(bowed_messages, messages['label'])
print(classification_report(messages['label'], naive_model.predict(bowed_messages)))

             precision    recall  f1-score   support

        ham       0.97      0.99      0.98       747
       spam       0.99      0.97      0.98       747

avg / total       0.98      0.98      0.98      1494



In [35]:
# CountVectorizer и токенизация с сохранением знаков
bow = CountVectorizer(analyzer=tokenize_punc)
bow.fit_transform(messages['message'])
bowed_messages = bow.transform(messages['message'])
naive_model = MultinomialNB()
naive_model.fit(bowed_messages, messages['label'])
print(classification_report(messages['label'], naive_model.predict(bowed_messages)))

             precision    recall  f1-score   support

        ham       0.98      1.00      0.99       747
       spam       1.00      0.98      0.99       747

avg / total       0.99      0.99      0.99      1494



In [36]:
# TfidfVectorizer и токенизация с удалением знаков
bow = TfidfVectorizer(analyzer=tokenize)
bow.fit_transform(messages['message'])
bowed_messages = bow.transform(messages['message'])
naive_model = MultinomialNB()
naive_model.fit(bowed_messages, messages['label'])
print(classification_report(messages['label'], naive_model.predict(bowed_messages)))

             precision    recall  f1-score   support

        ham       0.97      0.99      0.98       747
       spam       0.99      0.97      0.98       747

avg / total       0.98      0.98      0.98      1494



In [37]:
# TfidfVectorizer и токенизация с сохранением знаков
bow = TfidfVectorizer(analyzer=tokenize_punc)
bow.fit_transform(messages['message'])
bowed_messages = bow.transform(messages['message'])
naive_model = MultinomialNB()
naive_model.fit(bowed_messages, messages['label'])
print(classification_report(messages['label'], naive_model.predict(bowed_messages)))

             precision    recall  f1-score   support

        ham       0.98      1.00      0.99       747
       spam       1.00      0.97      0.99       747

avg / total       0.99      0.99      0.99      1494



Проемжуточные результаты: кажется TfidfVectorizer и CountVectorizer дают одинаковые результаты, тогда я буду использовать только второе.
Определение спама с сохранением пунктуации как-то неправдопобоно высоко. Мне кажется, это из-за того, что в спаме очень часто повторяются одни и те же слова в конце предложения, после которых идёт !, например, я видела в нашей дате: cash! ну и так далее

In [45]:
# CountVectorizer и токенизация с удалением знаков и стемматизацией
bow = CountVectorizer(analyzer=tokenize_stem)
bow.fit_transform(messages['message'])
bowed_messages = bow.transform(messages['message'])
naive_model = MultinomialNB()
naive_model.fit(bowed_messages, messages['label'])
print(classification_report(messages['label'], naive_model.predict(bowed_messages)))

             precision    recall  f1-score   support

        ham       0.97      0.99      0.98       747
       spam       0.99      0.97      0.98       747

avg / total       0.98      0.98      0.98      1494



In [42]:
# CountVectorizer и токенизация с удалением знаков, стемматизацией и стоп-словами
bow = CountVectorizer(analyzer=tokenize, stop_words=stopwords)
bow.fit_transform(messages['message'])
bowed_messages = bow.transform(messages['message'])
naive_model = MultinomialNB()
naive_model.fit(bowed_messages, messages['label'])
print(classification_report(messages['label'], naive_model.predict(bowed_messages)))

             precision    recall  f1-score   support

        ham       0.97      0.99      0.98       747
       spam       0.99      0.97      0.98       747

avg / total       0.98      0.98      0.98      1494



In [ ]:
Мне кажется, я что-то делаю не так, не может быть такого, чтобы у меня получались одинаковые результаты с применением разных параметров